src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -174,6 +169,36 @@ entails SolverOptions{..} constraint context hints =
     forClassName ctx cn@(Qualified (Just mn) _) tys = concatMap (findDicts ctx cn) (ordNub (Nothing : Just mn : map Just (mapMaybe ctorModules tys)))
     forClassName _ _ _ = internalError "forClassName: expected qualified class name"
 
+    -- Right biased union of two row types
+    unionRows :: Type -> Type -> Type -> Maybe (Type, Maybe [Constraint])
+    unionRows l r o =
+        let (ls, lEnd) = rowToList l
+            (rs, rEnd) = rowToList r
+            (os, oEnd) = rowToList o
+            lm = M.fromList ls
+            rm = M.fromList rs
+            lrUnion = M.toList $ M.union rm lm
+            unioned = rowFromList (lrUnion, oEnd)
+            constraintDiff a b m = Just $ [ Constraint C.RowUnion [a, b, rowFromList (M.toList $ M.difference (M.fromList os) m, oEnd)] Nothing ]
+        in case (isClosedRow lEnd, isClosedRow rEnd) of
```

Comment:
@paf31 Yeah I didn't think what I was doing was quite correct, I'm still very much a noob as far as compiling haskell like functional languages, so it might not be correct (I didn't know if checking equality of the tail ending with == is enough for example..)

I will have another crack doing it the way you've described.. So the union is a simple matter of appending the known labels (do I need to remove exact duplicates?) The good thing is that I have a decent set of test cases now.